docs-client/src/containers/MethodPage/GraphqlRequestBody.tsx

```diff
@@ -0,0 +1,190 @@
+    subField
+  }
+}`;
+const jsonPlaceHolder = jsonPrettify('{"foo":"bar"}');
+
+const QueryEditor = loadable(() => import('./QueryEditor'));
+
+interface Props {
+  requestBodyOpen: boolean;
+  requestBody: string;
+  onEditRequestBodyClick: React.Dispatch<unknown>;
+  onDebugFormChange: (value: string) => void;
+  schemaUrlPath: string;
+}
+
+const toggle = (prev: boolean, override: unknown) => {
+  if (typeof override === 'boolean') {
+    return override;
+  }
+  return !prev;
+};
+
+const parseJson = (requestBody: string) => {
+  let parsedJson;
+  try {
+    parsedJson = JSON.parse(requestBody);
+  } catch (e) {
+    // ignored
+  }
+  return parsedJson;
+};
+
+const GraphqlRequestBody: React.FunctionComponent<Props> = ({
+  requestBodyOpen,
+  requestBody,
+  onEditRequestBodyClick,
+  onDebugFormChange,
+  schemaUrlPath,
+}) => {
+  const [queryOpen, toggleQueryOpen] = useReducer(toggle, true);
+  const [variablesOpen, toggleVariablesOpen] = useReducer(toggle, false);
+
+  const [query, setQuery] = useState('');
+  const [variables, setVariables] = useState('');
+  const [schema, setSchema] = useState<GraphQLSchema | undefined>();
+
+  useEffect(() => {
+    (async () => {
+      const headers: any = {
+        'Content-Type': 'application/json',
+        Accept: 'application/json',
+      };
+      if (process.env.WEBPACK_DEV === 'true') {
+        headers[docServiceDebug] = 'true';
+      }
+      const httpResponse = await fetch(schemaUrlPath, {
+        method: 'POST',
+        headers,
+
+        body: JSON.stringify({
+          operationName: 'IntrospectionQuery',
+          // See https://github.com/graphql/graphiql/blob/8ac05f8b141b6f5cb4449c62ad67a34115490ac8/packages/graphiql/src/utility/introspectionQueries.ts#L16...L22
+          query: getIntrospectionQuery().replace(
+            'subscriptionType { name }',
+            '',
+          ),
+        }),
+      });
+      const result = await httpResponse.json();
+      if (typeof result !== 'string' && 'data' in result) {
+        setSchema(buildClientSchema(result.data));
+      }
+    })();
+  }, [schemaUrlPath]);
+
+  useEffect(() => {
+    const parsedJson = parseJson(requestBody);
+    if (parsedJson === undefined) {
+      return;
+    }
+
+    setQuery(parsedJson.query);
+    if (typeof parsedJson.variables === 'object') {
+      setVariables(jsonPrettify(JSON.stringify(parsedJson.variables)));
+    } else {
+      setVariables(parsedJson.variables);
+    }
+  }, [requestBody]);
+
+  useEffect(() => {
+    const parsedVariables = parseJson(variables);
+    if (parsedVariables === undefined) {
+      return;
+    }
+
+    onDebugFormChange(
+      JSON.stringify({
+        query,
+        variables,
+      }),
+    );
+  }, [onDebugFormChange, query, variables]);
+
+  const onQueryFromChange = useCallback((value) => {
+    setQuery(value);
+  }, []);
+
+  const onVariablesFromChange = useCallback((value) => {
+    setVariables(value);
+  }, []);
```

Comment:
nit; are these necessary if they don't have any special logic? Isn't it possible to just use `setQuery`/`setVariables` directly?

Reply:
I think it should have clear semantics for the behavior just like any other `useCallback`. 🤔

Reply:
I'm no react guru myself so I'm not sure 😅 
From what I understand using `useCallback` returns a memoized function but we're not actually memoizing anything... [ref](https://reactjs.org/docs/hooks-reference.html#usecallback)
Anyways, I think it's fine to leave as is 👍